lib/state-store.js

```diff
@@ -0,0 +1,311 @@
+      }
+    };
+
+    try {
+      await client.update(params);
+      logger.debug(`The consumer "${consumerId}" is now registered.`);
+    } catch (err) {
+      if (err.code !== 'ConditionalCheckFailedException') {
+        logger.error(err);
+        throw err;
+      }
+    }
+
+    params = {
+      Key: { consumerGroup, streamName },
+      UpdateExpression: 'set #a.#b.#c = :x',
+      ExpressionAttributeNames: {
+        '#a': 'consumers',
+        '#b': consumerId,
+        '#c': 'heartbeat'
+      },
+      ExpressionAttributeValues: {
+        ':x': heartbeat
+      }
+    };
+
+    await client.update(params);
+  }
+
+  async getShardAndStreamState(shardId, shardData) {
+    const privateProps = internal(this);
+    const { logger, client, consumerGroup, streamName } = privateProps;
+
+    let streamState = await getStreamState(this);
+    let shardState = streamState.shards[shardId];
+    if (shardState !== undefined) return { shardState, streamState };
+
+    const params = {
+      Key: { consumerGroup, streamName },
+      UpdateExpression: 'SET #a.#b = :x',
+      ConditionExpression: 'attribute_not_exists(#a.#b)',
+      ExpressionAttributeNames: {
+        '#a': 'shards',
+        '#b': shardId
+      },
+      ExpressionAttributeValues: {
+        ':x': {
+          ...shardData,
+          checkpoint: null,
+          depleted: false,
+          leaseExpiration: null,
+          leaseOwner: null,
+          version: generate()
+        }
+      }
+    };
+
+    try {
+      await client.update(params);
+    } catch (err) {
+      if (err.code !== 'ConditionalCheckFailedException') {
+        logger.error(err);
+        throw err;
+      }
+    }
+
+    streamState = await getStreamState(this);
+    shardState = streamState.shards[shardId];
+    return { shardState, streamState };
+  }
+
+  async lockShardLease(shardId, leaseTermTimeout, version) {
+    const privateProps = internal(this);
+    const { consumerId, client, consumerGroup, logger, streamName } = privateProps;
+
+    const params = {
+      Key: { consumerGroup, streamName },
+      UpdateExpression: 'SET #a.#b.#c = :w, #a.#b.#d = :x, #a.#b.#e = :y',
+      ConditionExpression: '#a.#b.#e = :z',
+      ExpressionAttributeNames: {
+        '#a': 'shards',
+        '#b': shardId,
+        '#c': 'leaseOwner',
+        '#d': 'leaseExpiration',
+        '#e': 'version'
+      },
+      ExpressionAttributeValues: {
+        ':w': consumerId,
+        ':x': new Date(Date.now() + leaseTermTimeout).toISOString(),
+        ':y': generate(),
+        ':z': version
+      }
+    };
+
+    try {
+      await client.update(params);
+      return true;
+    } catch (err) {
+      if (err.code !== 'ConditionalCheckFailedException') {
+        logger.error(err);
+        throw err;
+      }
+      return false;
+    }
+  }
+
+  async releaseShardLease(shardId, version) {
+    const privateProps = internal(this);
+    const { docClient, logger, streamName } = privateProps;
```

Comment:
Why are we using docClient here but client in the other functions?

Reply:
`client` is an instance of [AWS.DynamoDB](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/DynamoDB.html) while `docClient` is an instance of [AWS.DynamoDB.DocumentClient](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/DynamoDB/DocumentClient.html). The later one allows put/get calls using native JS types instead of the DynamoDB ones.